drivers: gpio: Add fxl6408 driver

Ilagas opened the pull request:
Add driver for the fxl6408 gpio-expander. The driver makes it possible to read and write to the gpio pins over I2C. See the README file of the sample application for general usage of the driver. The driver is inspired by the PCA95XX driver. Currently, the driver does not support interrupts or HIGH-Z mode which are options on the fxl6408. I'm not completely sure about HIGH-Z mode; should HIGH-Z mode be included in the driver and if so, how? In the gpio api, I could not find any support for HIGH-Z.

Signed-off-by: Imre Lagas Imre_Lagas@hotmail.com

galak commented on drivers/gpio/Kconfig.fxl6408:
```diff
@@ -0,0 +1,21 @@
+# ST Microelectronics STM32L151XC MCU
```
This comment about STM32L151XC seems out of place.

Ilagas replied:
I've removed this line

galak commented on dts/bindings/gpio/fcs,fxl6408.yaml:
```diff
@@ -0,0 +1,24 @@
+    label:
+      required: true
+
+    has-pud:
```
Where is this property being used in the code?

Ilagas replied:
I've removed the has-pud part since it wasn't used in code.

galak reviewed (commented):
Why do we need a unique sample for this driver?

galak reviewed (changes requested):
please rebase to remove the merge commit.

Ilagas commented:
@galak, I added the sample application mainly because i thought it would be useful for verifying that the driver is working. Should I remove the sample application? I've also pushed a new commit which I believe should resolve the merge commit.

Ilagas commented:
I have deleted the sample application since it didn't add that much and i had included it more or less to show that the driver is working. Just for completeness; I've tested all functions in the gpio API and they all work.

galak commented on dts/bindings/gpio/fcs,fxl6408.yaml:
```diff
@@ -0,0 +1,19 @@
+
+description: FXL6408 I2C-based GPIO expander
+
+compatible: "fcs,fxl6408"
```
Curious where `fcs` is coming from?

Ilagas replied:
I found it in /zephyr/dts/bindings/vendor-prefixes.txt. It stands for the vendor Fairchild Semiconductor.

galak reviewed (changes requested):
In general its best to merge commits and force push updates to your PR when you make changes so the commits are just the logical commits that would end up getting merged. 

So for example, since you are dropping the sample, the commit change that adds and and removes it should go away.

Additionally, all the other changes, dts, initial driver, and fixes for code compliance should get merged into one commit.

mnkp commented on drivers/gpio/gpio_fxl6408.c:
```diff
@@ -0,0 +1,436 @@
+	return ret;
+}
+
+int gpio_fxl6408_init(const struct device *device)
```
We traditionally keep _init function toward the bottom of the file. Just above the `struct` defining API functions. Let's move it there.

mnkp reviewed (commented):
This is a partial review.

> should HIGH-Z mode be included in the driver and if so, how? In the gpio api, I could not find any support for HIGH-Z.

The driver should configure pin in High-Z state if neither `GPIO_INPUT` nor `GPIO_OUTPUT` flag is provided.

Ilagas commented:
I've run into a problem. I have rebased to master, but now whenever I try to build I get the following warning:
```
/home/ilagas/zephyrproject/zephyr/drivers/gpio/gpio_fxl6408.c:436:13: warning: Macro is deprecated
  436 | DT_INST_FOREACH_STATUS_OKAY(GPIO_FXL6408_DEVICE_INSTANCE)
```
I've found this slack thread which states that it might be a bug: https://zephyrproject.slack.com/archives/C18PLHN8H/p1620392150084400

I dont really know where to start looking to fix this warning. Can I continue the PR or should this issue be fixed first?

mnkp commented:
> ```
> /home/ilagas/zephyrproject/zephyr/drivers/gpio/gpio_fxl6408.c:436:13: warning: Macro is deprecated
>   436 | DT_INST_FOREACH_STATUS_OKAY(GPIO_FXL6408_DEVICE_INSTANCE)
> ```

Unfortunately, the message is confusing but there is probably no way to avoid it. The power management has recently been reworked and `device_pm_control_nop` function was removed. You need to replace `device_pm_control_nop` with `NULL` in `DEVICE_DT_INST_DEFINE` macro.

Ilagas commented on dts/bindings/gpio/fcs,fxl6408.yaml:
```diff
@@ -0,0 +1,19 @@
+      required: true
+
+    "#gpio-cells":
+      const: 1
```
I've changed this to `const: 2`. I must have misinterpreted this value; what is this used for exactly?

mnkp replied:
This is indeed not very well documented in DT Specification. This is related to 
```
gpio-cells:
  - pin
  - flags
```
and tells how many parameters are passed after the gpio phandle in `*gpios` parameter, as in 
```
gpios = <&pioc 8 GPIO_ACTIVE_LOW>;
```

mnkp replied:
It looks like it has been reverted back to 1. Should be `const: 2`.

Ilagas commented on drivers/gpio/gpio_fxl6408.c:
```diff
@@ -0,0 +1,429 @@
+
+	ret = i2c_reg_write_byte(i2c_master, config->i2c_slave_addr, reg,
+					 value);
+	if (ret == 0) {
+		*cache = value;
+	} else {
+		LOG_ERR("error writing to register 0x%X (%d)", reg,
+			ret);
+	}
+	LOG_DBG("Write: REG[0x%X] = 0x%X", reg, *cache);
```
What do you mean by this exactly? Do you mean that `read_port_regs` returns ret if it fails? If so, should both return out of the function on fail or should neither?

Ilagas commented:
It seems I have broken things with my latest push. Ill fix it as soon as possible.

Ilagas commented:
I just noticed that the buildkite test has failed. When I look into it, it has problems with nxp,lpc-gpio.yaml. I did not change anything in this file and I don't get why this different board would be causing problems. Any ideas why this is happening?

PerMac commented:
> I just noticed that the buildkite test has failed. When I look into it, it has problems with nxp,lpc-gpio.yaml. I did not change anything in this file and I don't get why this different board would be causing problems. Any ideas why this is happening?

@Ilagas Check the "Files changed" by your PR. You have changed this file in fact. The buildkit error is not a CI issue.

PerMac commented on dts/bindings/gpio/nxp,lpc-gpio.yaml:
```diff
@@ -17,14 +17,6 @@ properties:
     "#gpio-cells":
       const: 2
 
-    port:
-      type: int
-      required: true
-      description: The GPIO port this node describes
-      enum:
-        - 0
-        - 1
-
```
@Ilagas I guess this relates to the error you see in buildkite

mnkp commented on drivers/gpio/gpio_fxl6408.c:
```diff
@@ -0,0 +1,438 @@
+	k_sem_take(&drv_data->lock, K_FOREVER);
+
+	ret = setup_pin_dir(dev, pin, flags);
+	if (ret) {
```
Let's write it as `(ret != 0)`, here and in a few more places, to use the same style. This is recommended (but not yet enforced) but the coding guidelines.

mnkp commented on drivers/gpio/gpio_fxl6408.c:
```diff
@@ -0,0 +1,438 @@
+	.pin_interrupt_configure = gpio_fxl6408_pin_interrupt_configure
+};
+
+#define GPIO_FXL6408_DEVICE_INSTANCE(inst)                                  \
+	static const struct gpio_fxl6408_config gpio_fxl6408_##inst##_cfg = {   \
+	.common = {                                                             \
+		.port_pin_mask = GPIO_PORT_PIN_MASK_FROM_DT_INST(inst),             \
+	},                                                                      \
+	.i2c_slave_addr = DT_INST_REG_ADDR(inst),                               \
+	.i2c_master = DEVICE_DT_GET(DT_INST_BUS(inst))                          \
+};                                                                          \
+\
+	static struct gpio_fxl6408_drv_data gpio_fxl6408_##inst##_drvdata = {   \
+		.reg_cache = {                                                      \
+			.input = 0x0,                                                   \
+			.output = 0xFF,                                                 \
+			.dir = 0x0,                                                     \
+			.high_z = 0x0,                                                  \
+			.pud_en = 0xFF,                                                 \
+			.pud_sel = 0xFF                                                 \
+		}                                                                   \
+	};                                                                      \
+\
+	DEVICE_DT_INST_DEFINE(inst, gpio_fxl6408_init, NULL,                    \
+				  &gpio_fxl6408_##inst##_drvdata,                           \
+				  &gpio_fxl6408_##inst##_cfg, POST_KERNEL,                  \
+				  CONFIG_GPIO_FXL6408_INIT_PRIORITY,                        \
+				  &gpio_fxl_driver);
```
Let's limit line length to 80 characters.

mnkp reviewed (commented):
LGTM with a few more, mainly cosmetic changes. @Ilagas could you please test the driver with `tests/drivers/gpio/gpio_api_1pin/` testcase.

mnkp reviewed (changes requested):
@Ilagas please confirm that tests/drivers/gpio/gpio_api_1pin/ testcase is passing.

Some CI checks are failing. Please clean it up.

Ilagas commented:
> @Ilagas please confirm that tests/drivers/gpio/gpio_api_1pin/ testcase is passing.
> 
> Some CI checks are failing. Please clean it up.

The gpio_api_1pin test runs succesfully. I've tested it on a stm32  with the  FXL6408.  I've used the following line to run the test:
`scripts/twister -vv --device-testing --device-serial /dev/ttyUSB0 -p stm32l1_disco -T tests/drivers/gpio/gpio_api_1pin/`

The twister.log file should be attached. 
[twister.log](https://github.com/zephyrproject-rtos/zephyr/files/6704262/twister.log)

Ilagas commented:
What does strike me as odd though is that when running the  tests/drivers/gpio/gpio_api_1pin/ testcase, I would assume the LEDs attached to the FXL6408 would be blinking. This is not the case. Is this a problem?

Ilagas commented:
> What does strike me as odd though is that when running the tests/drivers/gpio/gpio_api_1pin/ testcase, I would assume the LEDs attached to the FXL6408 would be blinking. This is not the case. Is this a problem?

Whenever I run the test, it doesn't test the gpio-expander but rather the GPIO pins of the stm32 itself. Any idea how I can configure the testcase to test the pins of the FXL6408?

Ilagas commented:
By the way, a while back I added accidentally comitted a few unrelated files which trigerred Github to add the NXP platfrom tag. This added a few additional reviewers to the PR. Should the tag and the reviewers be manually removed or does it not matter that much?

mnkp reviewed (commented):
> There seems to be a problem. Whenever I run the test, it doesn't test the gpio-expander but rather the GPIO pins of the stm32 itself. Any idea how I can configure the testcase to test the pins of the FXL6408? I need to change led0 in the device tree so that is uses a pin from the FXL6408, but I cant seem to make this work.

To test is quickly on stm32l1_disco board you can simply modify `gpios` property of the `green_led` / `led0` to use FXL6408 node that you already have. If unsure you can check `boards/` folder for existing examples. Other in tree GPIO-expanders are cy8c95xx, pca95xx, pcal6408a, sx1509b. If you grep board directory for those names you'll definitely find something.

> By the way, a while back I added accidentally comitted a few unrelated files which trigerred Github to add the NXP platfrom tag. This added a few additional reviewers to the PR. Should the tag and the reviewers be manually removed from the PR or does it not matter that much?

Normally you would remove reviewers that were added accidentally. In case of this PR, since there were few, you can probably leave it as is.

Ilagas commented:
> @Ilagas please confirm that tests/drivers/gpio/gpio_api_1pin/ testcase is passing.

I've tried running this test but many of the tests are failing because they require the pin to be configured as both input and output. I don't think it's possible on the fxl6408 to write to a pin and read it at the same time. I've come to this conclusion by trying to write/read a pin in a sample application. It also states in the datasheet that when a pin is configured as output the input status register bit of that pin has no effect (see table 10 in the datasheet). For now I'm removing the functionality to set both input and output for a pin but I think it would be wise of someone could doublecheck if this really is not possoible on the fxl6408.

Ilagas commented:
I've succesfully ran the tests/drivers/gpio/gpio_api_1pin/ testcase. Do note that many of the tests were simply skipped because setting input and ouput at the same time was not supported. 
[twister.log](https://github.com/zephyrproject-rtos/zephyr/files/6732992/twister.log)

mnkp commented on drivers/gpio/gpio_fxl6408.c:
```diff
@@ -0,0 +1,438 @@
+
+	return write_port_regs(dev, REG_PUD_EN, &drv_data->reg_cache.pud_en,
+				   value);
+}
```
Sorry, I put the comment in the wrong place. That was misleading. It's easier / cleaner to implement this check inside `write_port_regs` function.

mnkp commented:
> I've succesfully ran the tests/drivers/gpio/gpio_api_1pin/ testcase. Do note that many of the tests were simply skipped because setting input and ouput at the same time was not supported.

I double checked the datasheet and indeed in/out mode is not supported. The gpio_api_1pin testcase is not very useful in this case but at least the `gpio_pin_configure` function returns now the correct return value.

@Ilagas in that case could you test the driver with `tests/drivers/gpio/gpio_basic_api/` testcase. It's a bit more cumbersome to setup but you'll find multiple examples in the `tests/drivers/gpio/gpio_basic_api/boards/` directory.

@galak I'll need your help here. I assume that fxl6408 gpio-expander is not available on any of the evaluation boards. Do we want to have in-tree board overlay/config files to enable testing the driver on some arbitrary board with `tests/drivers/gpio/gpio_basic_api/` testcase or is it enough that we add an entry to `tests/drivers/build_all/gpio/` to make sure the driver builds?

Ilagas commented:
I finally managed to get the test `tests/drivers/gpio/gpio_basic_api/` up and running. I've used one of the supported boards; the `nucleo_g071rb`. The test is failing however because the test tries to set the `GPIO_ACTIVE_LOW` flag. This is currently not supported. I cant find any other implementation of the `GPIO_ACTIVE_LOW` flag and some other drivers don´t support it. If you would like to implement the `GPIO_ACTIVE_LOW` flag, what would be the best way to do this?

mnkp commented:
> The test is failing however because the test tries to set the `GPIO_ACTIVE_LOW` flag.

The flag is handled fully by the code in `include/drivers/gpio.h`. There is nothing driver needs to do apart from making sure that `struct gpio_driver_data common;` is the first field in the runtime data struct, which it is. The testcase must be failing for some other reason.

Ilagas committed:
drivers: gpio: Add fxl6408 driver

Add a driver for the fxl6408 gpio-expander using gpio api.

Origin: original

Signed-off-by: Imre Lagas <Imre_Lagas@hotmail.com>

Ilagas commented:
I have ran the `gpio_basic_api` testcast testcase partially succesful. The failure I encountered earlier was because the `GPIO_ACTIVE_LOW` flag was not in my list of supported flags. As far as I understand I can just add this flag to the list of supported flags so I've done this. 
The test that writes and reads the pins executes succesful and many of the tests to test callbacks are skipped because gpio interrupts are not supported by the driver. Only the last test fails because It tries to configure the pin with the flag `GPIO_INT_DEBOUNCE`. This is not supported by the driver and it's not in the list of supported flags. How should this be adressed?

[handler.log](https://github.com/zephyrproject-rtos/zephyr/files/6825595/handler.log)

Ilagas commented:
I just notices many of the buildkite tests have failed. This time I didn't accidentally include unrelated files or made changes in any unrelated files. Also, I performed a rebase to the main branch before pushing. I have no idea why these unrelated errors pop up. Any ideas what could be causing this?

mnkp commented:
> Only the last test fails because It tries to configure the pin with the flag `GPIO_INT_DEBOUNCE`. This is not supported by the driver and it's not in the list of supported flags. How should this be adressed?

Yes,  `GPIO_INT_DEBOUNCE` is a bit special. According to the documentation

```
 * @note Drivers that do not support a debounce feature should ignore
 * this flag rather than rejecting the configuration with -ENOTSUP.
```

So you need to add the flag to the list of supported flags. Even if it's not supported by the hardware.

mnkp commented:
@Ilagas there is one more change required. You'll need to add an entry to the `tests/drivers/build_all/gpio/` testcase to make sure the driver is build by the CI. It's not complicated. You can `git grep -i PCAL6408A tests/drivers/build_all/gpio/` to see what needs to be done. PCAL6408A is another GPIO expander.

Also, please rebase the PR on the latest master.

Ilagas commented:
i have added the testcase to `tests/drivers/build_all/gpio/` and added `GPIO_INT_DEBOUNCE` in the list of allowed tags. The test runs succesfully now:

[twister.log](https://github.com/zephyrproject-rtos/zephyr/files/6840479/twister.log)

mnkp reviewed (commented):
LGTM. For the build_all testcase you need to enable the driver by setting `CONFIG_GPIO_FXL6408=y` in `tests/drivers/build_all/gpio/prj.conf`.

Ilagas commented:
I've added `CONFIG_GPIO_FXL6408=y` to the prj.conf file. I have added it in the middle so it's underneath the `CONFIG_I2C` line and not underneath `CONFIG_SPI`. Are there any rules as to where to add lines or should I just always add lines to the end of a file?

carlescufi commented:
@Ilagas could you please rebsae this?

Ilagas commented:
I have rebased to the zephyr main branch

Ilagas commented:
Hi, do I need to do anything additional? If possible I would like to finish the PR soon.

PerMac commented:
This PR is out of my competence scope and I removed myself from the reviewers. Github shows that I had a comment, but it is no longer applicable as that was resolved.
